### ISSUE TITLE
Consider combinator that fails fast after first error is found

### ISSUE THREAD

**Issue by saraiva132** [zio/zio#3732]
After a discussion on discord regarding the current combinators for validation / accumulation of failures I realized we do not have one that accumulates successes until an error is found. I guess the opposite of `validateFirst`. 

Possible signature could be (names are suggestions):

```
def failFast[R, E, A, B](
    in: Iterable[A]
  )(f: A => ZIO[R, E, B])(implicit ev: CanFail[E]): ZIO[R, Nothing, (Option[E], List[B])]
```

or

```
def shortCircuit[R, E, A, B](
    in: Iterable[A]
  )(f: A => ZIO[R, E, B])(implicit ev: CanFail[E]): ZIO[R, Nothing, List[B]]
```

In the end, the first can be seen as a specialized version of partition. The second is just something of its own.. 

WDYT ?

**Comment by luis3m** [comment]
Maybe `validateAccum` or `validateFirstError` to keep it under the `validate` family

**Comment by saraiva132** [comment]
`validateAccum` would be nice if we do not care about the error thrown. Validate is already lossy to begin with, but we currently discard the successes and this quite different so it might create confusion. My first instinct would for it to be close to the partition variants but it is also a bit different from them.

**Comment by luis3m** [comment]
Agree, we need a fitting name then. IMO `failFast` doesn't say anything about the intent of the function.

**Comment by saraiva132** [comment]
@adamgfraser do you have an opinion on this?

**Comment by adamgfraser** [comment]
How about `collectWhileSuccess`?

**Comment by luis3m** [comment]
@adamgfraser nice, sounds good to me and there's a similar combinator in ZStream so would be consistent

**Comment by saraiva132** [comment]
I will take this one then, sounds fun
